### PR TITLE
Relay compatibility

### PR DESCRIPTION
**Pull request by tothandras**
- [x] node endpoint support
- [x] pagination support
image (using the example app)
- [x] GraphiQL example
- [x] Test data for the example
- [x] Tests
- [ ] `viewer` root

### CODE REVIEW

**Comment by burkhardr** [comment]
Hi @tothandras . Please take a look at https://github.com/RisingStack/graffiti-mongoose/issues/22 as well. What I thin might be missing from your list is:

* Mutations
* Fragment support
* Connections


Great to see you've made a start on this though!

File: src/type.js [hunk 1]
```diff
@@ -12,10 +13,30 @@ import {getField} from './field';
...
  */
 function getTypes (models) {
-  var types = reduce(models, (types, model) => {
+
+  var types = {};
+  var typeNames = Object.keys(models);
+  var {nodeInterface, nodeField} = nodeDefinitions(
+    (globalId) => {
+      var {type, id} = fromGlobalId(globalId);
+      if (typeNames.indexOf(type) > -1) {
```
Comment (burkhardr) [inline src/type.js]: Could this just be `if(models[type])`?

File: src/type.js [hunk 2]
```diff
@@ -12,10 +13,30 @@ import {getField} from './field';
...
+        return models[type].model.findById(id).then((mongooseObj) => {
+          var obj = mongooseObj.toObject();
+          obj._graphqlType = type;
+          return obj;
+        });
+      }
+      return null;
+    }, (obj) => {
+      return types[obj._graphqlType];
```
Comment (burkhardr) [inline src/type.js]: Is this WIP? `types` is empty at this point

**Comment by tothandras** [comment]
Hi @burkhardr, 
thanks for the review! I was just playing around with the current codebase to see where it has be changed. I tried to get more familiar with it by implementing a simple feature like supporting `node(id)` queries. More will be coming soon.

**Comment by parkan** [comment]
What's the rationale behind creating a new `GraphQLID` `id` instead of reusing `_id` (which is generally UUID-like and indexed already)? To avoid exposing machine or timing information?

File: src/type.js (hunk 2 above)
Comment (parkan) [inline src/type.js]: I'm guessing the thinking is that the type resolver function closes over `var types` so it's available (and filled) when it's invoked later. `nodeField` is not accessed until it's returned from this function, after `types` is populated.

File: src/type.js (hunk 1 above)
Comment (parkan) [inline src/type.js]: Maybe just `Object.hasOwnProperty(type)` and omit typeNames

File: src/type.js (hunk 2 above)
Comment (burkhardr) [inline src/type.js]: I completely missed L35! You're right

**Comment by tothandras** [comment]
I am almost finished. Please start reviewing as this contains a lot of change.

**Comment by tothandras** [comment]
I've included all my changes, please review @burkhardr, @parkan, @hekike!

**Comment by hardchor** [comment]
Will do tonight!

**Comment by hardchor** [comment]
Couple of tests failing for me mate, so won't even let me build it:

```
 1) e2e get schema "before all" hook for "should extract tree from paths":
     Error: Schema must contain unique named types but contains multiple types named "User".
      at invariant (node_modules/graphql/jsutils/invariant.js:20:11)
      at typeMapReducer (node_modules/graphql/type/schema.js:138:41)
      at Array.reduce (native)
      at typeMapReducer (node_modules/graphql/type/schema.js:146:44)
      at Array.reduce (native)
      at typeMapReducer (node_modules/graphql/type/schema.js:150:41)
      at node_modules/graphql/type/schema.js:164:22
      at Array.forEach (native)
      at typeMapReducer (node_modules/graphql/type/schema.js:155:30)
      at Array.reduce (native)
      at new GraphQLSchema (node_modules/graphql/type/schema.js:67:92)
      at getSchema (src/field/index.js:89:10)
      at Context.<anonymous> (src/e2e.spec.js:14:16)
      at Hook.Runnable.run (node_modules/co-mocha/lib/co-mocha.js:43:16)

  2) field getSchema should return a GraphQL schema:
     Error: Schema must contain unique named types but contains multiple types named "User".
      at invariant (node_modules/graphql/jsutils/invariant.js:20:11)
      at typeMapReducer (node_modules/graphql/type/schema.js:138:41)
      at Array.reduce (native)
      at typeMapReducer (node_modules/graphql/type/schema.js:146:44)
      at node_modules/graphql/type/schema.js:164:22
      at Array.forEach (native)
      at typeMapReducer (node_modules/graphql/type/schema.js:155:30)
      at Array.reduce (native)
      at new GraphQLSchema (node_modules/graphql/type/schema.js:67:92)
      at getSchema (src/field/index.js:89:10)
      at Context.getSchemaTest (src/field/field.spec.js:100:22)
      at Test.Runnable.run (node_modules/co-mocha/lib/co-mocha.js:43:16)
```

**Comment by hekike** [comment]
@hardchor it's working for me with `Node 4.1.0 `. Did you run `npm install`? Do you have running MongoDB?

**Comment by hardchor** [comment]
Mongo is running (the errors come from graphql anyway). Node is v4.1.2.

```bash
system in graffiti-mongoose/ on  
› node -v
v4.1.2
```

File: src/field/index.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        id: {
+          type: new GraphQLNonNull(GraphQLID),
+          description: `The ID of a ${name}`
+        }
+      },
+      resolve: getOneResolver(graffitiModel)
+    },
+    [pluralName]: {
+      type: new GraphQLList(type),
+      args: reduce(type._typeConfig.fields(), (args, field) => {
```
Comment (hardchor) [inline src/field/index.js]: I don't think relay likes root fields with `args` any other than `id`s. See https://github.com/facebook/relay/issues/112 for a discussion.

```
-- Relay Transform Error -- users.jsx --

Error: Invalid root field `users`; Relay only supports root fields with zero or one argument
```

**Comment by tothandras** [comment]
@hardchor I'm getting similar errors, when I call `getSchema` or `getTypes` multiple times in one application. But the tests should pass if you haven`t modified it.

File: src/field/index.js (hunk 3 above)
Comment (tothandras) [inline src/field/index.js]: Should I remove them?

**Comment by hardchor** [comment]
@tothandras Take a look at the link I posted. They propose introducing a root-level `viewer` field and then put everything else underneath that. What do you think?

**Comment by tothandras** [comment]
@hardchor Okay, I am going to add the `viewer` root endpoint and keep the rest as is.

**Comment by hardchor** [comment]
@tothandras Yes, and sadly I didn't manage to track down why. I've not modified anything, just used your PR directly, and had MongoDB running (which wasn't where the issue came from).

@hekike I don't think @tothandras had time to add the `viewer` root node by the time you'd merged. Should we open a new PR instead?

**Comment by hekike** [comment]
@tothandras  @hardchor yes please. I don't want to end up in an even bigger pr. ;)

**Comment by hardchor** [comment]
No worries ;) Could you tag the new release for npm, please?

**Comment by tothandras** [comment]
@hardchor the new version (3.0.0) is published to npm.

**Pull request merged**